expo-maps/android/src/main/java/expo/modules/maps/GoogleMapsView.kt

```diff
@@ -22,4 +23,16 @@ class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallbac
   override fun onMapReady(googleMap: GoogleMap) {
 
   }
+
+  fun mapView(mapType: MapType) {
+    val googleMapType = when (mapType) {
+      MapType.normal -> GoogleMap.MAP_TYPE_NORMAL
+      MapType.terrain -> GoogleMap.MAP_TYPE_TERRAIN
+      MapType.satellite -> GoogleMap.MAP_TYPE_SATELLITE
+      MapType.hybrid -> GoogleMap.MAP_TYPE_HYBRID
+    }
+    mapView.getMapAsync {map: GoogleMap ->
+      map.mapType = googleMapType
+    }
+  }
```

Comment:
Instead of calling getMapAsync again I would suggest saving GoogleMap object from onMapReady method to a class variable and modifying map via this object. If you try to do it you may come across a problem with onMapReady method being called after method which sets props (mapView in this case) and which needs initialized GoogleMap object. I think this might be solved by introducing and observing kotlin Flow variable which would emit a value when the map is ready. I would like to know if @lukmccall approves this approach. The code may look like this:

```suggestion

  class GoogleMapsView(context: Context): LinearLayout(context), OnMapReadyCallback {

  private val mapView: MapView = MapView(context)
  private lateinit var googleMap: GoogleMap
  private val mapReady = MutableStateFlow(false)

  val lifecycleEventListener = MapViewLifecycleEventListener(mapView)

  init {
    mapView.onCreate(null)
    mapView.getMapAsync(this)
    mapView.onStart()
    mapView.onResume()
    addView(mapView)
  }

  override fun onMapReady(googleMap: GoogleMap) {
    this.googleMap = googleMap
    CoroutineScope(Dispatchers.Default).launch {
      mapReady.emit(true)
    }
  }

  fun mapView(mapType: MapType) {
    val googleMapType = when (mapType) {
      MapType.normal -> GoogleMap.MAP_TYPE_NORMAL
      MapType.terain -> GoogleMap.MAP_TYPE_TERRAIN
      MapType.satelite -> GoogleMap.MAP_TYPE_SATELLITE
      MapType.hybrid -> GoogleMap.MAP_TYPE_HYBRID
    }

    updateMap {
      googleMap.mapType = googleMapType
    }
  }

  private fun updateMap(update: () -> Unit) {
    CoroutineScope(Dispatchers.Default).launch {
      mapReady.collectLatest {
        if (it) {
          withContext(Dispatchers.Main) {
            update()
          }
          cancel()
        }
      }
    }
  }
}
```
To use coroutines you need to add a new gradle dependency to a library level build.gradle file:
`implementation "org.jetbrains.kotlinx:kotlinx-coroutines-android:1.4.3"`

Besides that I think that the mapView method should be renamed to something like changeMapType or as it is done in swift - simply mapType

Reply:
Thanks @p-karcz for these very informative comments!